Feat/notifications 326

asoto-ideasdigitales opened the pull request:
Notificaciones a usuarios , envio de correos (Mail::queue)

Test 
<img width="667" height="238" alt="image" src="https://github.com/user-attachments/assets/02d17738-47a7-49fb-b001-f2986adc638d" />

asoto-ideasdigitales committed:
feat: Agrega controlador y lógica para el envío de notificaciones por correo

asoto-ideasdigitales committed:
refactor: Simplifica la ruta de notificaciones y mejora la autorización en el controlador de notificaciones

prodev443 commented on app/Http/Controllers/Api/NotificationController.php:
```diff
@@ -0,0 +1,42 @@
+
+class NotificationController extends Controller
+{
+    public function store(Request $request)
```
Utilicemos clases Request de validación para todos los endpoints que manejen entrada de datos por favor.

asoto-ideasdigitales replied:
Se agrega Request

prodev443 commented on routes/api.php:
```diff
@@ -228,6 +229,10 @@
 });
 
+Route::post('/notifications', [NotificationController::class, 'store'])
+    ->middleware(['auth:sanctum','role:admin|superadmin','permission:create-notifications'])
```
Dejemos de utilizar middleware de **role** en el archivo de rutas por favor

asoto-ideasdigitales replied:
Modificado

prodev443 commented on app/Http/Controllers/Api/NotificationController.php:
```diff
@@ -0,0 +1,42 @@
+        $customers = User::role('customer')->get();
+
+        // Enviar notificación por correo a cada customer
+        foreach ($customers as $customer) {
```
Disculpa que te cambie la jugada, me equivoqué.

Creo que encolar el envío de correo a cada cliente durante la solicitud no es la mejor opción.

Si mantenemos la lógica justo como la programaste, el escenario (con 300 clientes por ejemplo) será el siguiente:

1. Solicitud de envío de notificaciones al backend
2. **Backend ingresa 300 procesos en cola pendientes en la base de datos** _Aquí es donde veo el problema_
3. La solicitud termina con una respuesta exitosa

Quizás una opción más óptima sea encolar solo un proceso que en **segundo plano** se encargue de efectuar el envío a cada **cliente**, **el proceso ya no tendría que encolar los envíos**, solo hay que garantizar que **pueda tener una ejecución prolongada** (investigar si es necesario configurar php de alguna forma para ello).

Si no te convence esta última alternativa lo vemos durante la daily por favor.

prodev443 commented on resources/views/emails/notification.blade.php:
¿Validaste esto con front?

Si no lo hiciste es necesario que lo hagas en una reunión o a través de una discusión (puedes usar la que ya está de notificaciones), por favor.

prodev443 commented on tests/Feature/NotificationControllerTest.php:
```diff
@@ -0,0 +1,190 @@
+    
+    $this->admin = User::factory()->create();
+    $this->admin->assignRole('admin');
+    $this->admin->givePermissionTo('create-notifications');
```
No es necesario que asignes el rol admin si estás asignando el permiso explícitamente

asoto-ideasdigitales replied:
Corregido

prodev443 commented on tests/Feature/NotificationControllerTest.php:
```diff
@@ -0,0 +1,190 @@
+    $this->admin->givePermissionTo('create-notifications');
+    
+    $this->customer1 = User::factory()->create();
+    $this->customer1->assignRole('customer');
```
Asignar el rol **customer** está bien porque justamente el requerimiento dice que a los **clientes** deben llegarles las notificaciones

asoto-ideasdigitales replied:
ok

prodev443 commented on tests/Feature/NotificationControllerTest.php:
```diff
@@ -0,0 +1,190 @@
+
+        it('should allow access to store with permission', function () {
+            Mail::fake();
+            \Spatie\Permission\Models\Permission::firstOrCreate(['name' => 'create-notifications']);
```
La siembra de los permisos se ejecuta para todos los test automáticamente, esta línea no sería necesaria

prodev443 commented on tests/Feature/NotificationControllerTest.php:
```diff
@@ -0,0 +1,190 @@
+            
+            $response->assertStatus(201)
+                ->assertJsonStructure([
+                    'message',
```
El retorno de la estructura con **message** y **data** no tiene mucho sentido, más que nada porque los mensajes no contienen información muy relevante y el front no los está utilizando.

Mientras mandemos el código correcto, creo que estamos siendo lo suficientemente descriptivos con el resultado de la operación.

Te sugiero que retornes directamente el arreglo que estás colocando en data (sin anidarlo):

```php
return [
     'title',
    'message',
    'recipients_count',
    'created_at'
];
```

asoto-ideasdigitales replied:
Modificado

prodev443 reviewed (changes requested):
Favor de atender los ajustes solicitados

asoto-ideasdigitales committed:
feat: Implementa el envío de notificaciones en segundo plano y mejora la validación de solicitudes

asoto-ideasdigitales committed:
feat: Ajusta la gestión de la fecha de creación en las notificaciones y mejora la validación de la solicitud

asoto-ideasdigitales committed:
Implementar sistema de notificaciones push usando Firebase Cloud Messaging

Se agregó el campo fcm_token a la tabla users mediante una migración.

Se creó la clase PushNotification (notification) para manejar mensajes de FCM.

Se desarrolló el job SendPushNotification para enviar notificaciones a usuarios con tokens FCM válidos.

Se actualizó NotificationControllerTest para probar notificaciones push en lugar de notificaciones por correo.

Se eliminaron el código y las aserciones relacionadas con correo electrónico en las pruebas.

Se añadió un mock de Firebase Messaging en las pruebas para simular el envío de notificaciones.

asoto-ideasdigitales committed:
Merge branch 'main' into feat/notifications-326

asoto-ideasdigitales committed:
feat: Mejora la implementación de notificaciones push y actualiza las pruebas para incluir tokens FCM

asoto-ideasdigitales committed:
Merge branch 'main' into feat/notifications-326

asoto-ideasdigitales committed:
feat(Firebase): agrega configuración y credenciales para Firebase

asoto-ideasdigitales committed:
feat(Firebase): agrega archivo de credenciales de Firebase

asoto-ideasdigitales committed:
feat(Firebase): elimina archivo de credenciales de Firebase de ejemplo

prodev443 reviewed (changes requested):
> [!IMPORTANT]
> **Malas noticias**
> No va a ser posible realizar el despliegue colocando las credenciales de firebase manualmente en el storage, **Laravel Cloud** no proporciona acceso al almacenamiento local.

## Requested changes

- [x] Crear la ruta `PUT /api/firebase/config` en la que se deben enviar todos los atributos como **application/json**, para posteriormente procesarlos y almacenarlos como fichero json en el almacenamiento local.
- [x] Permiso para proteger la ruta: **update-system-config**

## DoD

- [x] Controlador específico para esta operación
- [x] Test que verifique validaciones y lógica
- [x] Request para aislar la validación de datos

## Notes

_Hay que actualizar la rama **feat/notifications-326** con **main**, está atrasada_

asoto-ideasdigitales committed:
Merge branch 'main' into feat/notifications-326

asoto-ideasdigitales committed:
feat(Firebase): implementar Firebase configuration update endpoint y validacion
